NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +52,224 @@ function Base.show(io::IO, s::CategoryProduct)
   return print(io, ")")
 end
 
-category_show(io::IO, k, v) = print(io, v)
-
+category_show(io::IO, ::Int, v) = print(io, v)
 category_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return categories_isless(categories(s1), categories(s2))
+end
+function Base.isless(s1::CategoryProduct, s2::CategoryProduct)
+  return categories_isless(categories(s1), categories(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for CategoryProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+# get clean results when mixing implementations
+categories_isequal(::Tuple, ::NamedTuple) = false
+categories_isequal(::NamedTuple, ::Tuple) = false
+
+categories_isless(::NamedTuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(::Tuple, ::NamedTuple) = throw(ArgumentError("Not implemented"))
+
+categories_type(::Type{<:CategoryProduct{T}}) where {T} = T
+
+function categories_fusion_rule(cats1, cats2)
+  diff_cat = CategoryProduct(categories_diff(cats1, cats2))
+  shared1, shared2 = categories_common(cats1, cats2)
+  fused = map(fusion_rule, values(shared1), values(shared2))
```

Comment:
A potential issue I see here is that in the `NamedTuple` version, `categories_common` is defined like this:
```julia
function categories_common(nt1::NamedTuple, nt2::NamedTuple)
  return intersect_keys(nt1, nt2), intersect_keys(nt2, nt1)
end
```
That doesn't appear to impose a (sorted) order on the keys of `shared1` and `shared2`, and therefore `values` may not be in the correct order. Doesn't a sorting of the keys need to happen in `categories_common` for this to be correct if the keys of `cats1` and `cats2` are not sorted? Or is the assumption that `cats1` and `cats2` already arrive sorted? If so, I think it would be good to put a comment about that or maybe even a check somewhere, I could imagine that causing issues if we extend this to other data structures like `CategoryProduct{<:Dict}` in the future.